### PR TITLE
fix(cli): add hierarchical Escape navigation

### LINKED ISSUE #9588: CLI workflow child focus needs an explicit Esc-back path
## Problem

Selecting a workflow child stream can feel like a modal dead end. The active transcript changes to the child, but there is no persistent parent breadcrumb or obvious way to return to the workflow root.

Current controls are easy to misread:

- `Tab` transfers focus to/from the child list rather than navigating to the parent.
- Bare `Esc` is owned by run interruption or the Esc+1..9 focus chord.
- The footer can show `Esc 1..9 focus`, but does not expose `Esc back` while a child is active.
- Selecting a child blurs the list, removing the only visible root row from keyboard focus.

This was reproduced while inspecting a running workflow script in the CLI TUI.

## Current code seam

- `packages/cli/src/chat/tui/App.tsx:550-599` handles child-list focus, Tab, meta-focus chords, and bare-Escape interruption. There is no child-to-parent navigation branch.
- `packages/cli/src/chat/tui/panes/statusBarDisplay.ts` already knows `isChildStream`, but normal shortcut bindings do not turn that fact into an `Esc back` hint.
- `packages/cli/src/chat/tui/panes/SubagentList.tsx:393-414` focuses selected streams, while cancel only returns keyboard ownership to input.

## Expected behavior

Navigation should be hierarchical and non-modal:

1. From a focused child, bare `Esc` returns to its immediate parent.
2. From a nested grandchild, repeated `Esc` walks one parent at a time.
3. At the root, existing run-interruption behavior remains unchanged.
4. Foreground modal/form/list cancellation keeps precedence.
5. The footer explicitly displays `Esc back` whenever that action is available.
6. Esc+1..9/Alt+1..9 focus chords remain usable without making bare Escape undiscoverable.

## Acceptance tests

- child → immediate parent
- nested child → immediate parent, not root
- root Escape interruption unchanged
- foreground/list Escape precedence unchanged
- status-bar binding says `Esc back` only when applicable
- terminal mode using Esc as the numeric-focus modifier remains covered

### PR DESCRIPTION
**Pull request by LionSR**
## Summary

- make bare `Esc` move from a focused workflow child to its immediate parent, one level at a time
- preserve root interruption, foreground/list cancellation precedence, and Esc/Alt numeric focus chords
- clamp child-list and input-history arrows instead of implicitly transferring keyboard ownership
- show `Esc back` in the footer exactly when parent navigation is available

## Verification

- focused CLI navigation/input/status/resolver/session suite: 359 tests passed
- review follow-up suite: 126 tests passed
- `npm run typecheck:cli`
- `npm run typecheck:test-kernel`
- `npm run lint`
- CLI architecture and React compiler smoke checks
- Prettier and `git diff --check`

Closes #9588

---

> [!NOTE]
> **Medium Risk**
> Central `App` keyboard routing for Escape, interrupts, and focus changed with many edge cases; regressions could affect run cancellation vs navigation, though coverage is extensive in integration tests.
> 
> **Overview**
> **Bare `Esc` now walks up the stream tree** one parent at a time (via `focusStream` / approval promotion) before falling back to the existing run-interrupt behavior on the root. Deferred-Esc handling records the parent at schedule time so delayed or “failed chord” keys don’t navigate after topology, foreground panes, or lifecycle focus changes; printable characters typed during that window can be **`appendInput`**’d when parent back re-enables the prompt.
> 
> **Arrow keys no longer implicitly move focus** between the prompt, history, and child list: `InputBar` drops `onFocusChildList` and clamps history ↑/↓; `SubagentList` removes boundary `onCancel` so ↑/↓ clamp on rows and **Tab** is the explicit path back to input (still owned in `App`).
> 
> The status footer gains **`Esc back`** when the active stream has a parent, with width-aware binding rows so the hint appears only when it fits.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7faa34e12ce1eb09de9d2d59ffb87db6e797eb78. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/cli/src/chat/tui/App.tsx [hunk 1]
```diff
@@ -480,11 +481,20 @@ export function App(props: AppProps): React.JSX.Element {
...
 
-  const scheduleEscapeInterrupt = (streamId: StreamTabId) => {
+  const handleBareEscape = (streamId: StreamTabId): void => {
+    const parentId = parentStreamSignal.get().get(streamId);
+    if (parentId !== undefined) {
+      focusStreamAndPromoteApprovals(parentId);
+      return;
+    }
+    triggerEscapeInterrupt(escapeInterruptStateRef.current, streamId);
+  };
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/tui/App.tsx:518]: Fixed in ea92a5d508. Deferred bare Escape now revalidates current App-level ownership through the existing policy before parent focus; regression tests cover a foreground pane appearing during the chord window.

File: packages/cli/src/chat/tui/App.tsx [hunk 2]
```diff
@@ -504,10 +514,7 @@ export function App(props: AppProps): React.JSX.Element {
         input.length > 0
       ) {
         if (handleMetaShortcut(input)) return;
-        triggerEscapeInterrupt(
-          escapeInterruptStateRef.current,
-          pendingEscape.streamId,
-        );
+        handleBareEscape(pendingEscape.streamId);
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/tui/App.tsx]: Fixed in ea92a5d508. A second quick bare Escape resolves the pending action and preserves the second action against the newly focused stream; the grandchild-to-child-to-root sequence is covered behaviorally.

File: packages/cli/src/chat/tui/App.tsx [hunk 3]
```diff
@@ -480,11 +481,20 @@ export function App(props: AppProps): React.JSX.Element {
       });
   };
 
-  const scheduleEscapeInterrupt = (streamId: StreamTabId) => {
+  const handleBareEscape = (streamId: StreamTabId): void => {
+    const parentId = parentStreamSignal.get().get(streamId);
+    if (parentId !== undefined) {
+      focusStreamAndPromoteApprovals(parentId);
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/tui/App.tsx:514]: Fixed in ea92a5d508. Both timer and failed-chord paths recheck current foreground/input/overlay ownership before navigating.

File: packages/cli/src/chat/tui/panes/statusBarDisplay.ts [hunk 4]
```diff
@@ -602,7 +609,12 @@ function statusBarBindingsText(
     transcriptAvailable && statusBarBindingRow([childList, fullOutput, ctrlC]),
     childNavigationAvailable && childList,
     transcriptAvailable && statusBarBindingRow([fullOutput, ctrlC]),
-  ];
+  ].map((candidate) =>
+    parentBack && candidate
+      ? statusBarBindingRow([parentBack, candidate])
+      : candidate,
+  );
+  if (parentBack) candidates.push(parentBack);
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/tui/panes/statusBarDisplay.ts:628]: Fixed in ea92a5d508. The width cascade now tries `Esc back · Ctrl-C ...` before child-only or Esc-only fallbacks, with an exact moderate-width regression assertion.

File: packages/cli/src/chat/tui/App.tsx [hunk 5]
```diff
@@ -497,17 +522,30 @@ export function App(props: AppProps): React.JSX.Element {
...
+        if (
+          shouldDeferEscapeInterruptForMetaChord({
+            shortcutModifierLabel: defaultShortcutModifierLabel(),
+            streamFocusAvailable: sessionViews.length > 0,
+          })
+        ) {
+          scheduleBareEscape(currentStreamId);
+        } else {
+          handleBareEscape(currentStreamId);
+        }
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/tui/App.tsx:569]: Fixed in 55020d0332. A second quick Escape chains only when resolving the first action actually changes the active stream. Nested back navigation still advances twice, while a root interruption now fires exactly once; both are behaviorally tested beyond the chord window.

File: packages/cli/src/chat/tui/App.tsx [hunk 6]
```diff
@@ -497,17 +522,30 @@ export function App(props: AppProps): React.JSX.Element {
...
+        return;
+      }
+      if (!key.ctrl && !key.tab && input.length > 0) {
         if (handleMetaShortcut(input)) return;
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/tui/App.tsx]: Fixed in a11cdf44e9. Pending Esc-prefix digits now recheck current Escape ownership before focusing; a foreground surface appearing during the chord window keeps focus and remains visible. Normal Esc+digit and ordinary Alt+digit paths remain covered and functional.

File: packages/cli/src/chat/tui/App.tsx [hunk 7]
```diff
@@ -480,11 +481,40 @@ export function App(props: AppProps): React.JSX.Element {
...
+    }
+    return triggerEscapeInterrupt(escapeInterruptStateRef.current, streamId);
+  };
+
+  const scheduleBareEscape = (streamId: StreamTabId) => {
     clearPendingEscapeInterrupt();
     const timer = setTimeout(() => {
       pendingEscapeInterrupt.current = undefined;
-      triggerEscapeInterrupt(escapeInterruptStateRef.current, streamId);
+      handleBareEscape(streamId);
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/tui/App.tsx]: Fixed in 76e621f74a. Deferred and failed-chord bare-Escape actions now resolve only while their captured stream is still active, so lifecycle-driven focus advancement cannot be rolled back. Timer and failed-chord races are covered; double-Esc and normal chords remain green.

File: packages/cli/src/chat/tui/App.tsx [hunk 8]
```diff
@@ -497,17 +532,32 @@ export function App(props: AppProps): React.JSX.Element {
...
+        ) {
+          scheduleBareEscape(currentStreamId);
+        } else {
+          handleBareEscape(currentStreamId);
+        }
+        return;
+      }
+      if (!key.ctrl && !key.tab && input.length > 0) {
+        if (appOwnsEscape() && handleMetaShortcut(input)) return;
+        handleBareEscape(pendingEscape.streamId);
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/tui/App.tsx]: Fixed in 023d3196bc. When a failed Escape chord moves from a completed child with disabled input to its enabled parent, printable input is appended to the parent exactly once. Regression tests cover both disabled and already-enabled child input.

File: packages/cli/src/chat/tui/App.tsx [hunk 9]
```diff
@@ -480,11 +481,45 @@ export function App(props: AppProps): React.JSX.Element {
...
+      !bareEscapeActive(streamId)
+    ) {
+      return false;
+    }
+    const parentId = parentStreamSignal.get().get(streamId);
+    if (parentId !== undefined) {
+      focusStreamAndPromoteApprovals(parentId);
+      return true;
+    }
+    return triggerEscapeInterrupt(escapeInterruptStateRef.current, streamId);
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/tui/App.tsx:517]: Fixed in 023d3196bc. A deferred Escape captures the parent relationship present when it is scheduled and is discarded if that topology changes, so promotion or detachment cannot turn a back action into an interrupt. Timer and failed-chord promotion races are both tested.

File: packages/cli/src/chat/tui/panes/statusBarDisplay.ts [hunk 10]
```diff
@@ -600,9 +607,15 @@ function statusBarBindingsText(
     childNavigationAvailable &&
       statusBarBindingRow([childList, fullOutput, ctrlC]),
     transcriptAvailable && statusBarBindingRow([childList, fullOutput, ctrlC]),
+    parentNavigationAvailable && ctrlC,
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/tui/panes/statusBarDisplay.ts:620]: Fixed in 023d3196bc. The width cascade now tries Esc back + Ctrl-T full output + Ctrl-C before the parent Ctrl-C-only fallback. An exact moderate-width regression assertion verifies that the transcript binding is retained.

File: packages/cli/src/chat/tui/App.tsx [hunk 11]
```diff
@@ -497,17 +543,51 @@ export function App(props: AppProps): React.JSX.Element {
...
+          pendingEscape.parentStreamId,
         );
+        const printableInput =
+          !key.meta &&
+          !key.return &&
+          metaChordInput(input, key) === undefined &&
+          [...input].every((character) => !isUnhandledControlInput(character));
+        if (handled && inputWasDisabled && printableInput) {
+          inputBarRef.current?.appendInput(input);
+        }
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/tui/App.tsx:589]: Fixed through 79e276a9b3. Empty-input arrow events now resolve the captured deferred back action instead of cancelling it. All four arrow sequences are covered from a completed child, and the tests also verify that no empty prompt is submitted.

[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/tui/App.tsx:589]: Fixed in 5482a580eb and covered comprehensively in 79e276a9b3. Pending bare Escape now treats all four parsed arrow flags as failed-chord resolution even when `input` is empty; arrows are excluded from printable replay. Regression coverage exercises Up/Down/Left/Right from a disabled completed child and verifies no interrupt or submit, while the existing no-pending input/list test now covers all four directions. Focused CLI interaction suites pass (144 tests), along with lint and both CLI/test-kernel typechecks.

File: packages/cli/src/chat/tui/App.tsx [hunk 12]
```diff
@@ -497,17 +543,54 @@ export function App(props: AppProps): React.JSX.Element {
...
-          escapeInterruptStateRef.current,
+      if (isEscapeInput(input, key)) {
+        const previousStreamId = activeStreamIdSignal.get();
+        if (
+          !handlePendingBareEscape(
+            pendingEscape.streamId,
+            pendingEscape.parentStreamId,
+          )
+        ) {
+          return;
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/tui/App.tsx]: Fixed in d4abb7d797. A second bare Escape is now reinterpreted against the current stream when the deferred action became stale, while the existing same-stream suppression remains in force after a successful pending action. Regression coverage includes both focus advancement and parent-topology invalidation; the focused suite passes 24/24.

[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/tui/App.tsx]: Fixed in d4abb7d797 (current integrated head 7faa34e12c). A second bare Escape now retries against the current active stream when the prior deferred action is rejected, while successful pending actions retain the existing quick-double-Escape/root single-interrupt behavior. Behavioral regressions cover both lifecycle focus advancement (fresh root interrupt) and parent-topology invalidation (fresh promoted-stream interrupt). The focused CLI suites pass 147 tests, with Prettier, lint, CLI typecheck, and test-kernel typecheck also passing after merging current origin/main.
